runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/WindmillStreamSender.java

```diff
@@ -0,0 +1,157 @@
+  private final Supplier<GetDataStream> getDataStream;
+  private final Supplier<CommitWorkStream> commitWorkStream;
+  private final StreamingEngineThrottleTimers streamingEngineThrottleTimers;
+
+  private WindmillStreamSender(
+      CloudWindmillServiceV1Alpha1Stub stub,
+      GetWorkRequest getWorkRequest,
+      AtomicReference<GetWorkBudget> getWorkBudget,
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      ProcessWorkItem processWorkItem) {
+    this.started = new AtomicBoolean(false);
+    this.getWorkBudget = getWorkBudget;
+    this.streamingEngineThrottleTimers = StreamingEngineThrottleTimers.create();
+
+    // All streams are memoized/cached since they are expensive to create and some implementations
+    // perform side effects on construction (i.e. sending initial requests to the stream server to
+    // initiate the streaming RPC connection). Stream instances connect/reconnect internally so we
+    // can reuse the same instance through the entire lifecycle of WindmillStreamSender.
+    this.getDataStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createGetDataStream(
+                    stub, streamingEngineThrottleTimers.getDataThrottleTimer()));
+    this.commitWorkStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createCommitWorkStream(
+                    stub, streamingEngineThrottleTimers.commitWorkThrottleTimer()));
+    this.getWorkStream =
+        Suppliers.memoize(
+            () ->
+                streamingEngineStreamFactory.createDirectGetWorkStream(
+                    stub,
+                    withRequestBudget(getWorkRequest, getWorkBudget.get()),
+                    streamingEngineThrottleTimers.getWorkThrottleTimer(),
+                    getDataStream,
+                    commitWorkStream,
+                    processWorkItem));
+  }
+
+  public static WindmillStreamSender create(
+      CloudWindmillServiceV1Alpha1Stub stub,
+      GetWorkRequest getWorkRequest,
+      GetWorkBudget getWorkBudget,
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      ProcessWorkItem processWorkItem) {
+    return new WindmillStreamSender(
+        stub,
+        getWorkRequest,
+        new AtomicReference<>(getWorkBudget),
+        streamingEngineStreamFactory,
+        processWorkItem);
+  }
+
+  private static GetWorkRequest withRequestBudget(GetWorkRequest request, GetWorkBudget budget) {
+    return request.toBuilder().setMaxItems(budget.items()).setMaxBytes(budget.bytes()).build();
+  }
+
+  @SuppressWarnings("ReturnValueIgnored")
+  public void startStreams() {
+    Preconditions.checkState(
+        !getWorkBudget.get().equals(GetWorkBudget.noBudget()), "Cannot GetWork with no budget.");
+    getWorkStream.get();
+    getDataStream.get();
+    commitWorkStream.get();
+    // *stream.get() is all memoized in a threadsafe manner.
+    started.compareAndSet(false, true);
```

Comment:
no need to compare, just unconditionally set true

Reply:
done.